### ISSUE TITLE
FSDP models are not being correctly converted to another data-type after initialization.

### ISSUE THREAD

**Issue by ysiraichi** [pytorch/pytorch#137522]
### 🐛 Describe the bug

The test `test_to_float64_after_init` starts failing after introducing a dtype mismatch check for `lerp`. 

https://github.com/pytorch/pytorch/blob/d34b617bb98cc18a42e0af4df8e275f40c25dca9/test/distributed/_composable/fsdp/test_fully_shard_training.py#L202

**Context:** `lerp` eager implementation actually errors on mismatched data-types for its inputs. However, [its decomposition](https://github.com/pytorch/pytorch/blob/5e3e1c01515bac92bbe303d3fdbfee571949a96f/torch/_refs/__init__.py#L5074-L5080) implements type-promotion, instead. After fixing this issue (#136909), the above-mentioned test starts failing.

https://github.com/pytorch/pytorch/blob/b499083a91e9233e7e9292a4f898ef6f74d8bdc6/aten/src/ATen/native/Lerp.cpp#L15-L20

Upon closer inspection, I observed that: even though the `DTensor` had the same dtype for both `start` and `end`, when it got to the decomposition (without #136909), the dtypes were different. Moving the dtype conversion statement in the test `model.to(dtype)` before calling `fully_sharded` on the model fixes the issue (defeats the purpose of the test, though).

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @albanD

### Versions

PyTorch version: 2.6.0a0+git85eed60
Is debug build: True
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

**Comment by albanD** [comment]
tentative high priority as it is blocking adding completely valid meta implementations!

**Comment by awgu** [comment]
I can take a look at this tomorrow morning.

**Comment by awgu** [comment]
Is this a `DTensor` issue? I checked that parameter `DTensor`, parameter local tensor, gradient `DTensor`, and gradient local tensor are all fp64.
```
diff --git a/test/distributed/_composable/fsdp/test_fully_shard_training.py b/test/distributed/_composable/fsdp/test_fully_shard_training.py
index de89b5c65e0..01f335ba91c 100644
--- a/test/distributed/_composable/fsdp/test_fully_shard_training.py
+++ b/test/distributed/_composable/fsdp/test_fully_shard_training.py
@@ -216,6 +216,7 @@ class TestFullyShardCastAfterInit(FSDPTestMultiThread):
         model.to(dtype)
         for param in model.parameters():
             self.assertEqual(param.dtype, dtype)
+            self.assertEqual(param.to_local().dtype, dtype)
         optim = torch.optim.Adam(model.parameters(), lr=1e-2, foreach=True)
         check_sharded_parity(self, ref_model, model)
         torch.manual_seed(42 + self.rank + 1)
@@ -227,6 +228,11 @@ class TestFullyShardCastAfterInit(FSDPTestMultiThread):
                 losses[-1].backward()
             self.assertEqual(losses[0], losses[1])
             check_sharded_parity(self, ref_model, model)
+            for param in model.parameters():
+                self.assertEqual(param.dtype, dtype)
+                self.assertEqual(param.to_local().dtype, dtype)
+                self.assertEqual(param.grad.dtype, dtype)
+                self.assertEqual(param.grad.to_local().dtype, dtype)
             for _optim in (ref_optim, optim):
                 _optim.step()
                 _optim.zero_grad(set_to_none=(iter_idx % 2 == 0))
```

**Comment by awgu** [comment]
Patching https://github.com/pytorch/pytorch/pull/136909 locally, I cannot repro the test failure 🤔

**Comment by awgu** [comment]
My machine died, so I will have to revisit this later.

I think as a next step, we should determine if it is related to DTensor specifically or if FSDP2 is related in anyway by seeing if we can repro the issue with fp64 DTensors and lerp.

**Comment by awgu** [comment]
@ysiraichi I think the issue should be fixed now.
